Loads from stack slots weren't being correctly marked with a def flag.

a1k0n opened the pull request:
This caused all kinds of warnings and occasional crashes with -O2+.

Commit:
Loads from stack slots weren't being correctly marked with a def flag.

This caused all kinds of warnings and occasional crashes with -O2+.

a1k0n commented:
One other issue that the machine code verifier chokes on (that's how I found this) is SP is not in the GR16 register class, so all the MOV SP, x instructions are technically invalid.  We should include SP in the register class and mark it reserved.  Then the verification passes will work.

This patch is enough to fix the crashing issues I've seen though.

a1k0n commented on lib/Target/DCPU16/DCPU16InstrInfo.cpp:
```diff
@@ -72,12 +72,25 @@ void DCPU16InstrInfo::loadRegFromStackSlot(MachineBasicBlock &MBB,
   if (RC == &DCPU16::GR16RegClass || RC == &DCPU16::GEXR16RegClass)
-    BuildMI(MBB, MI, DL, get(DCPU16::MOV16rm))
-      .addReg(DestReg).addFrameIndex(FrameIdx).addImm(0).addMemOperand(MMO);
+    BuildMI(MBB, MI, DL, get(DCPU16::MOV16rm), DestReg)
```
The last arg of the 5-arg form of BuildMI is the destination register, which gets implicitly marked as a def.  I could also have just added the Defined flag to the addReg(DestReg), but this is apparently more idiomatic.

krasin commented:
Generally, looks good.
It would be nice if you add one or two ex-crashers as test cases (you can specify -O2 in the first line of each test, which contains a command line to invoke.

krasin commented:
And thank you for spotting this issue. It does not look as an easy food.

a1k0n commented:
I will try.  The problem with stuff like this is it only crashes on sufficiently complicated code, so making a test case is hard.  However, it's easy to spot by using llc -verify-machineinstrs -- except that will still fail until we fix some SP related issues, as I mentioned above.  But I could at least write a test that says llc -verify-machineinstrs doesn't generate an "Explicit definition marked as use" failure.  I'll do that.

krasin commented:
Sounds good (llc -verify-machineinstrs). Or, if you have a C file that leads to the crash (even if it's large enough), you might want to post it here, I will spend some time on this weekend and try to make a minimal crasher.

Commit:
Add def-use test via -verify-machineinstrs and a complex expression

It has to be complex enough to require stack spills to exercise this
particular issue.

Commit:
Add SP to the GR16 register class.

This is OK as it's already marked as a reserved reg by
DCPU16InstrInfo.cpp.  And it makes -validate-machineinstrs happy because
SP is a perfectly valid target operand for SET and ALU ops.

a1k0n commented:
Added tests, and fixed the SP issue.  That was easy.

krasin commented:
Awesome, LGTM
